pkg/volume/azure_dd/attacher.go

```diff
@@ -0,0 +1,291 @@
+	}
+
+	lun, err := strconv.Atoi(lunStr)
+	if err != nil {
+		return "", fmt.Errorf("WaitForAttach: wrong lun %q", lunStr)
+	}
+	ticker := time.NewTicker(checkSleepDuration)
+	defer ticker.Stop()
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			glog.V(5).Infof("Checking Azure disk %q(lun %s) is attached.", volumeSource.DiskName, lunStr)
+			scsiHostRescan(&osIOHandler{})
+			devicePath := findDiskByLun(lun, &osIOHandler{})
+			if devicePath != "" {
+				glog.Infof("Successfully found attached Azure disk %q(lun %s, device path %s).", volumeSource.DiskName, lunStr, devicePath)
+				return devicePath, nil
+			} else {
+				//Log error, if any, and continue checking periodically
+				glog.Errorf("Error Stat Azure disk (%q) is attached: %v", volumeSource.DiskName, err)
+			}
+		case <-timer.C:
+			return "", fmt.Errorf("Could not find attached Azure disk %q. Timeout waiting for mount paths to be created.", volumeSource.DiskName)
+		}
+	}
+}
+
+func (attacher *azureDiskAttacher) GetDeviceMountPath(
+	spec *volume.Spec) (string, error) {
+	volumeSource, err := getVolumeSource(spec)
+	if err != nil {
+		return "", err
+	}
+
+	return makeGlobalPDPath(attacher.host, volumeSource.DiskName), nil
+
+}
+
+func (attacher *azureDiskAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMountPath string) error {
+	mounter := attacher.host.GetMounter()
+	notMnt, err := mounter.IsLikelyNotMountPoint(deviceMountPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(deviceMountPath, 0750); err != nil {
+				return err
+			}
+			notMnt = true
+		} else {
+			return err
+		}
+	}
+
+	volumeSource, err := getVolumeSource(spec)
+	if err != nil {
+		return err
+	}
+
+	options := []string{}
+	if spec.ReadOnly {
+		options = append(options, "ro")
+	}
+	if notMnt {
+		diskMounter := &mount.SafeFormatAndMount{Interface: mounter, Runner: exec.New()}
+		err = diskMounter.FormatAndMount(devicePath, deviceMountPath, volumeSource.FSType, options)
+		if err != nil {
+			os.Remove(deviceMountPath)
+			return err
+		}
+	}
+	return nil
+}
+
+type azureDiskDetacher struct {
+	mounter mount.Interface
+	manager azureManager
+}
+
+var _ volume.Detacher = &azureDiskDetacher{}
+
+func (plugin *azureDataDiskPlugin) NewDetacher() (volume.Detacher, error) {
+	azure, err := getAzureDiskManager(plugin.host.GetCloudProvider())
+	if err != nil {
+		return nil, err
+	}
+
+	return &azureDiskDetacher{
+		mounter: plugin.host.GetMounter(),
+		manager: azure,
+	}, nil
+}
+
+func (detacher *azureDiskDetacher) Detach(dev string, hostName string) error {
+	if dev == "" {
+		return fmt.Errorf("invalid dev to detach: %q", dev)
+	}
+	instanceid, err := detacher.manager.InstanceID(hostName)
+	if err != nil {
+		return fmt.Errorf("failed to get azure instance id for host %q", hostName)
+	}
+	if ind := strings.LastIndex(instanceid, "/"); ind >= 0 {
+		instanceid = instanceid[(ind + 1):]
+	}
+	attachMutex.LockKey(instanceid)
```

Comment:
I think the lock here is no longer needed. Since v1.3.3, attach/unattach operatinos are invoked through operation_executor (k8s.io/kubernetes/pkg/volume/util/operationexecutor) which handles syntonization problem among different operations on the same volume.